presto-parquet/src/main/java/io/prestosql/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -78,32 +92,80 @@ protected boolean isValueNull()
+
+        if (prestoType == BooleanType.BOOLEAN && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.BOOLEAN) {
+            return new BooleanColumnReader(descriptor);
+        }
+
+        if (prestoType == TinyintType.TINYINT
+                || prestoType == SmallintType.SMALLINT
+                || prestoType == IntegerType.INTEGER
+                || prestoType == BigintType.BIGINT) {
+            // TODO verify no truncation occurs
+            if (parquetDecimalType.isPresent() && parquetDecimalType.get().getScale() == 0) {
+                // TODO use IntColumnReader or LongColumnReader when applicable
+                return decimalColumnReader.orElseThrow(() -> new VerifyException("decimalColumnReader not present when parquetDecimalType is present"));
+            }
+
+            if (!parquetDecimalType.isPresent()) {
+                if (descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT32) {
+                    return new IntColumnReader(descriptor);
+                }
+                if (descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT64) {
+                    return new LongColumnReader(descriptor);
                 }
-                return createDecimalColumnReader(descriptor).orElse(new LongColumnReader(descriptor));
-            case INT96:
-                return new TimestampColumnReader(descriptor);
-            case FLOAT:
-                return new FloatColumnReader(descriptor);
-            case DOUBLE:
-                return new DoubleColumnReader(descriptor);
-            case BINARY:
-                return createDecimalColumnReader(descriptor).orElse(new BinaryColumnReader(descriptor));
-            case FIXED_LEN_BYTE_ARRAY:
-                return createDecimalColumnReader(descriptor)
-                        .orElseThrow(() -> new PrestoException(NOT_SUPPORTED, " type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType()));
-            default:
-                throw new PrestoException(NOT_SUPPORTED, "Unsupported parquet type: " + descriptor.getPrimitiveType().getPrimitiveTypeName());
+            }
+        }
+
+        if (prestoType instanceof DecimalType) {
+            if (parquetDecimalType.isPresent() && ((DecimalType) prestoType).getScale() == parquetDecimalType.get().getScale()) {
+                // TODO allow rescaling
+                return decimalColumnReader.orElseThrow(() -> new VerifyException("decimalColumnReader not present when parquetDecimalType is present"));
+            }
+        }
+
+        if (prestoType == RealType.REAL && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.FLOAT) {
+            return new FloatColumnReader(descriptor);
+        }
+
+        if (prestoType == DoubleType.DOUBLE && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.DOUBLE) {
```

Comment:
could we read from parquet double to presto float previously?

Reply:
> could we read from parquet double to presto float previously?

no